forward43/api_ulule.py

```diff
@@ -0,0 +1,110 @@
+# -*- coding: utf-8 -*-
+"""
+This module uses the API for the platform Ulule to extract data on projects.
+
+Created on Wed Dec 16 22:06:55 2020
+@author: Marijke Thijssen
+"""
+# %%   Import packages
+import json
+import math
+# import pandas as pd
+import requests
+
+from support import timer
+
+# %%   Set variables
+URL = 'https://api.ulule.com/v1/'
+ENDPOINT_PROJECTS = 'search/projects/'
+QUERY = ['status:currently']  
+CRITERIA = ['id', 'name', 'description', 'status', 'type', 'country', 'location', 'absolute_url']
+
+# - id
+# Title - name
+# Description - description
+# Status - status
+# Innovation type - type
+# Country - country
+# City - location
+# Link - absolute_url
+
+# The description and name are only available if you get the project on ID. 
+# You also have to know the language you want the description in because the fields
+# have the country code appended to it, eg. decription_de, description_en.
+
+# Type: 1 (presale) or 2 (project)
+
+# Country is given as a two-letter ISO code
+
+# %%   Functions
+@timer
+def get_object(url):
+    """ Retrieve a response. """
+    response = requests.get(url)
+    if response.status_code != 200:
+        raise Exception(f'Unsuccessful GET on {url}: {response.status_code}')
+    
+    return response
+
+
+@timer
+def get_page(url, query, limit, page=1):
+    """ Retrieve a paginated response. """
+    response = requests.get(url, params={'q': '+'.join(query), 'offset': page * limit})
+    if response.status_code != 200:
+        raise Exception(f'Unsuccessful GET on {url}: {response.status_code}')
+    
+    return response
+
+
+def write_to_file(projects, file_name):
+    """ Write a list of projects to file. """
+    # project_df = pd.DataFrame(projects)
+    # project_df.to_csv(f'{file_name}.csv', index=False)
+    
+    with open(f'{file_name}.json', 'w', encoding='utf-8') as dest:
+        json.dump(projects, dest, ensure_ascii=False, indent=4)
+    
+    print(f'Wrote data to file: {file_name}.csv')
+
+
+@timer
+def main():
+    # Get initial response to assess number of projects
+    response = get_page(URL + ENDPOINT_PROJECTS, QUERY, limit=0)
+
+    data = response.json()
+    meta = data['meta']
+    n_projects = meta['total_count']
+    limit = meta['limit']
+    n_pages = math.ceil(n_projects / limit)
+    
+    #   Get list of IDs
+    id_list = []
+    for n in range(n_pages):
+        page = get_page(URL + ENDPOINT_PROJECTS, QUERY, limit, page=n).json()
+        projects = page['projects']
+        for p in projects:
+            id_list.append(p['id'])
+    
+    #   Get projects with criteria
+    project_list = []
+    for project_id in id_list:
+        project = get_object(URL + f'projects/{project_id}').json()
+        project = {k: project[k] for c in CRITERIA for k in project.keys() 
+                   if k.startswith(c) and len(k) < 15}                          # There are multiple 'description_x' fields; this selects only the desired one
```

Comment:
Wouldn't this give you all the `description_x` descriptions, and put them in the json, as they all start with description and have 14 chars given the 2 char country code?

If so: we'd need to map the `description_en` one to the base description so it conforms to the ES mapping and is searchable. 
If all projects have an en description we could add `description_en` to `CRITERIA` and rename the `description_en` key, probs something like:
```
project["description"] = project.pop("description_en")
```

Reply:
For both description and name all possibilities are retrieved. The comment in the code was actually to make sure that the `description_COUNTRYCODE` was selected instead of  `description_funding` or `description_yourself`. The problem with description (and name) is that the `description_en` is not always given. It could be that the description is given on only Spanish for example. For now, I'll only include projects with an English description and title. See `a263c48`.